crates/redeem/src/lib.rs

```diff
@@ -501,23 +502,48 @@ impl<T: Config> Pallet<T> {
             Error::<T>::AmountBelowDustAmount
         );
 
-        // vault will get rid of the btc + btc_inclusion_fee
-        ext::vault_registry::try_increase_to_be_redeemed_tokens::<T>(&vault_id, &vault_to_be_burned_tokens)?;
-
-        // lock full amount (inc. fee)
-        amount_wrapped.lock_on(&redeemer)?;
-        let redeem_id = ext::security::get_secure_id::<T>(&redeemer);
-
         let below_premium_redeem = ext::vault_registry::is_vault_below_premium_threshold::<T>(&vault_id)?;
         let currency_id = vault_id.collateral_currency();
 
         let premium_collateral = if below_premium_redeem {
-            let redeem_amount_wrapped_in_collateral = user_to_be_received_btc.convert_to(currency_id)?;
-            ext::fee::get_premium_redeem_fee::<T>(&redeem_amount_wrapped_in_collateral)?
+            // Calculate the secure vault capacity
+            let secure_vault_capacity = ext::vault_registry::vault_capacity_at_secure_threshold(&vault_id)?;
+            let to_be_backed_tokens = ext::vault_registry::vault_to_be_backed_tokens(&vault_id)?;
+            let difference_in_tokens_to_reach_secure_threshold =
+                to_be_backed_tokens.saturating_sub(&secure_vault_capacity)?;
+
+            if difference_in_tokens_to_reach_secure_threshold.gt(&user_to_be_received_btc)? {
+                let premium_tokens_in_collateral = user_to_be_received_btc.convert_to(currency_id)?;
+
+                ext::fee::get_premium_redeem_fee::<T>(&premium_tokens_in_collateral)?
+            } else {
+                // Formula = max_premium_collateral = (FEE * (oldTokens * EXCH * SECURE - oldCol)) / (SECURE - FEE)
```

Comment:
```suggestion
                // The goal of premium redeems is to get the vault back the a healthy collateralization ratio. As such, 
                // we only award a premium for the amount of tokens required to get the vault back to secure threshold.
                // The CollateralizationRate is defined as `totalCollateral / convertToCollateral(totalTokens)`
                // When paying a premium, the collateralization rate gets updated according to the following formula:
                //     `NewCollateralization = (oldCol - awardedPremium) / ( oldTokens*EXCH - awardedPremium/FEE)`
                // To calculate the maximum premium we are willing to pay, we set the newCollateralization to
                // the secure threshold, which gives:
                //     `SECURE = (oldCol - awardedPremium) / (oldTokens*EXCH - awardedPremium/FEE)``
                // We can rewrite this formula to calculate the `premium` amount that would get us to the secure threshold:
                //     `maxPremium = (oldTokens * EXCH * SECURE - oldCol) * (FEE / (SECURE - FEE))`
                // Which can be interpreted as:
                //     `maxPremium = missingCollateral * (FEE / (SECURE - FEE))
                // Note that to prevent repeated premium redeems while waiting for execution, we use to_be_backed_tokens
                // for `oldCol`, which takes into account pending issues and redeems
```

Can you add a comment like this please

Reply:
Note: in the comment I changed the formula compared to the one I gave earlier in the PR. 

Old: `maxPremium = (FEE * (oldCol - oldTokens * EXCH * SECURE)) / (FEE -SECURE)`
New: `maxPremium = (oldTokens * EXCH * SECURE - oldCol) * (FEE / (SECURE - FEE))`

This will result in positive intermediate results, and requires only a single multiplication between amounts and fixedpoint numbers